src/test/regress/sql/schema_based_sharding.sql

```diff
@@ -0,0 +1,412 @@
+ROLLBACK;
+
+SELECT COUNT(*)=0 FROM pg_dist_tenant_schema WHERE schemaid::regnamespace::text = 'tenant_8';
+SELECT COUNT(*)=0 FROM pg_dist_partition WHERE logicalrelid::text LIKE 'tenant_8.%';
+
+-- Verify that citus.enable_schema_based_sharding and citus.use_citus_managed_tables
+-- GUC don't interfere with each other when creating a table in tenant schema.
+--
+-- In utility hook, we check whether the CREATE TABLE command is issued on a tenant
+-- schema before checking whether citus.use_citus_managed_tables is set to ON to
+-- avoid converting the table into a Citus managed table unnecessarily.
+--
+-- If the CREATE TABLE command is issued on a tenant schema, we skip the check
+-- for citus.use_citus_managed_tables.
+SET citus.use_citus_managed_tables TO ON;
+CREATE TABLE tenant_7.tbl_3(a int, b text, PRIMARY KEY(a));
+RESET citus.use_citus_managed_tables;
+
+-- Verify that we don't unnecessarily convert a table into a Citus managed
+-- table when creating it with a pre-defined foreign key to a reference table.
+CREATE TABLE reference_table(a int PRIMARY KEY);
+SELECT create_reference_table('reference_table');
+
+-- Notice that tenant_7.tbl_4 have foreign keys both to tenant_7.tbl_3 and
+-- to reference_table.
+CREATE TABLE tenant_7.tbl_4(a int REFERENCES reference_table, FOREIGN KEY(a) REFERENCES tenant_7.tbl_3(a));
+
+SELECT COUNT(*)=2 FROM pg_dist_partition
+WHERE logicalrelid IN ('tenant_7.tbl_3'::regclass, 'tenant_7.tbl_4'::regclass) AND
+      partmethod = 'n' AND repmodel = 's' AND
+      colocationid = (SELECT colocationid FROM pg_dist_partition WHERE logicalrelid = 'tenant_7.tbl_1'::regclass);
+
+CREATE TABLE local_table(a int PRIMARY KEY);
+
+-- fails because tenant tables cannot have foreign keys to local tables
+CREATE TABLE tenant_7.tbl_5(a int REFERENCES local_table(a));
+
+-- Fails because tenant tables cannot have foreign keys to tenant tables
+-- that belong to different tenant schemas.
+CREATE TABLE tenant_5.tbl_5(a int, b text, FOREIGN KEY(a) REFERENCES tenant_7.tbl_3(a));
+
+\c - - - :worker_1_port
+
+-- test creating a tenant table from workers
+CREATE TABLE tenant_3.tbl_1(a int, b text);
+
+-- test creating a tenant schema from workers
+SET citus.enable_schema_based_sharding TO ON;
+CREATE SCHEMA worker_tenant_schema;
+SET citus.enable_schema_based_sharding TO OFF;
+
+-- Enable the GUC on workers to make sure that the CREATE SCHEMA/ TABLE
+-- commands that we send to workers don't recursively try creating a
+-- tenant schema / table.
+ALTER SYSTEM SET citus.enable_schema_based_sharding TO ON;
+SELECT pg_reload_conf();
+
+\c - - - :worker_2_port
+
+ALTER SYSTEM SET citus.enable_schema_based_sharding TO ON;
+SELECT pg_reload_conf();
+
+\c - - - :master_port
+
+SET search_path TO regular_schema;
+SET citus.next_shard_id TO 1940000;
+SET citus.shard_count TO 32;
+SET citus.shard_replication_factor TO 1;
+SET client_min_messages TO NOTICE;
+
+CREATE TABLE tenant_3.tbl_1(a int, b text);
```

Comment:
would be good to have a few search_path switching tests, especially with prepared statements and maybe functions, since that's a pattern that's quite unique to schema-based sharding.